pkg/engine/common.go

```diff
@@ -0,0 +1,62 @@
+package engine
+
+import (
+	"strings"
+
+	kubeutils "github.com/kyverno/kyverno/pkg/utils/kube"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// GetSubresourceGVKToAPIResourceMap returns a map of subresource GVK to APIResource. This is used to determine if a resource is a subresource.
+func GetSubresourceGVKToAPIResourceMap(kindsInPolicy []string, ctx *PolicyContext) map[string]*metav1.APIResource {
+	subresourceGVKToAPIResource := make(map[string]*metav1.APIResource)
+	for _, gvk := range kindsInPolicy {
+		gv, k := kubeutils.GetKindFromGVK(gvk)
+		parentKind, subresource := kubeutils.SplitSubresource(k)
+		// Len of subresources is non zero only when validation request was sent from CLI without connecting to the cluster.
+		if len(ctx.SubresourcesInPolicy) != 0 {
+			if subresource != "" {
+				for _, subresourceInPolicy := range ctx.SubresourcesInPolicy {
+					parentResourceGroupVersion := metav1.GroupVersion{
+						Group:   subresourceInPolicy.ParentResource.Group,
+						Version: subresourceInPolicy.ParentResource.Version,
+					}.String()
+					if gv == "" || kubeutils.GroupVersionMatches(gv, parentResourceGroupVersion) {
+						if parentKind == subresourceInPolicy.ParentResource.Kind {
+							if strings.ToLower(subresource) == strings.Split(subresourceInPolicy.APIResource.Name, "/")[1] {
+								subresourceGVKToAPIResource[gvk] = &(subresourceInPolicy.APIResource)
```

Comment:
Would be good to extract it into a util func.

Reply:
The util function would have a signature like:
```go
func funcName(gv string, parentResourceGroupVersion string, parentKind string, subresourceInPolicy struct {
	APIResource    metav1.APIResource
	ParentResource metav1.APIResource
}, subresource string) bool
```
Or something similar to it, I don't think it would make a good util function with so many parameters plus the `if` condition is only used twice, don't think extracting to a function would be necessary.